File: Answer-for-leetcode/C++/9.PalindromeNumber.cpp
/**************************************************************
9.回文数：palindrome-number
leetcode链接：https://leetcode-cn.com/problems/palindrome-number/
时间复杂度：O(log10(n))
空间复杂度：O(1)
**************************************************************/

#include<iostream>
using namespace std;

class Solution {
  public:
    bool palindromeNumber(int n) {
        //首先排除特殊情况
        if(n<0 || n%10==0 && n!=0) {
            return false;
        }
        //用于存储反转一半的数字
        int revertedNumber=0;
        //如果n有偶数位，则恰好反转一半后即n==revertedNumber时退出循环

        //如果n为奇数位，则反转至n<revertedNumber时退出循环
        //此时原数的中位数为revertedNumber的个位数
        while(n>revertedNumber) {
            revertedNumber = revertedNumber*10 + n%10;
            n/=10;
        }

        return n == revertedNumber || n == revertedNumber/10;
    }
};

int main() {
    int n;
    while(cin>>n) {
        Solution s;
        bool res=s.palindromeNumber(n);
        cout<<res<<endl;
    }
    return 0;
}

File: 三、Answer-for-leetcode/Python/9.PalindromeNumber.py
'''
9.回文数：palindrome-number
leetcode链接：https://leetcode-cn.com/problems/palindrome-number/
时间复杂度：O(log10(n))
空间复杂度：O(1)
'''


class Solution:
    def isPalindrome(self, x):
        '''
        :type x: int
        :return type:bool
        '''
        if x < 0:
            return False
        else:
            # [::-1]效果为：将字符串反转
            y = str(x)[::-1]
            if y == str(x):
                return True
            else:
                return False


if __name__ == "__main__":
    while True:
        x = input()
        x = int(x)
        s = Solution()
        res = s.isPalindrome(x)
        print(res)
